File: src/main/java/gargoyle/ct/prop/impl/CTBaseObservableProperty.java
package gargoyle.ct.prop.impl;

import gargoyle.ct.pref.CTPropertyChangeEvent;
import gargoyle.ct.pref.CTPropertyChangeListener;
import gargoyle.ct.prop.CTObservableProperty;
import gargoyle.ct.prop.CTProperty;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.function.Function;

public abstract class CTBaseObservableProperty<T> extends CTBaseProperty<T> implements CTObservableProperty<T> {
    protected CTBaseObservableProperty(Class<T> type, String name) {
        super(type, name);
    }

    @Nullable
    protected Thread firePropertyChange(T value, T oldValue) {
        if (Objects.equals(oldValue, value)) {
            return null;
        }
        return CTPropertyChangeManager.getInstance()
                .firePropertyChange(this,
                        new CTPropertyChangeEvent<>(this, name(), oldValue, value));
    }

    @Override
    public <T2> void bindBi(@NotNull CTObservableProperty<T2> property, @NotNull Function<T, T2> mapper, Function<T2, T> mapper2) {
        bind(property, mapper);
        property.bind(this, mapper2);
    }

    @SuppressWarnings("Convert2Lambda")
    @Override
    public <T2> void bind(@NotNull CTProperty<T2> property, @NotNull Function<T, T2> mapper) {
        T myValue = get();
        T2 otherValue = property.get();
        T2 newOtherValue = mapper.apply(myValue);
        if (!Objects.equals(otherValue, newOtherValue)) {
            property.set(newOtherValue);
        }
        addPropertyChangeListener(new CTPropertyChangeListener<T>() {
            @Override
            public void onPropertyChange(@NotNull CTPropertyChangeEvent<T> event) {
                onLateBind(property, mapper, event.getNewValue());
            }
        });
    }

    <T2> void onLateBind(@NotNull CTProperty<T2> property, @NotNull Function<T, T2> mapper, T newValue) {
        property.set(mapper.apply(newValue));
    }

    @Override
    public void bindBi(@NotNull CTObservableProperty<T> property) {
        bind(property);
        property.bind(this);
    }

    @SuppressWarnings("Convert2Lambda")
    @Override
    public void bind(@NotNull CTProperty<T> property) {
        T myValue = get();
        T otherValue = property.get();
        if (!Objects.equals(myValue, otherValue)) {
            onLateBind(property, myValue);
        }
        addPropertyChangeListener(new CTPropertyChangeListener<T>() {
            @Override
            public void onPropertyChange(@NotNull CTPropertyChangeEvent<T> event) {
                onLateBind(property, event.getNewValue());
            }
        });
    }

    void onLateBind(@NotNull CTProperty<T> property, T newValue) {
        property.set(newValue);
    }

    @Override
    public void addPropertyChangeListener(CTPropertyChangeListener<T> listener) {
        CTPropertyChangeManager.getInstance().addPropertyChangeListener(this, listener);
    }

    @Override
    public void removePropertyChangeListener(CTPropertyChangeListener<T> listener) {
        CTPropertyChangeManager.getInstance().removePropertyChangeListener(this, listener);
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/prop/impl/CTBaseObservableProperty.kt
package gargoyle.ct.util.prop.impl

import gargoyle.ct.util.pref.CTPropertyChangeEvent
import gargoyle.ct.util.pref.CTPropertyChangeListener
import gargoyle.ct.util.prop.CTObservableProperty
import gargoyle.ct.util.prop.CTProperty
import java.util.function.Function
import kotlin.reflect.KClass

abstract class CTBaseObservableProperty<T : Any> protected constructor(type: KClass<T>, name: String) :
    CTBaseProperty<T>(type, name), CTObservableProperty<T> {
    protected fun firePropertyChange(value: T, oldValue: T?): Thread? {
        return if (oldValue == value) {
            null
        } else CTPropertyChangeManager.instance
            .firePropertyChange(
                this,
                CTPropertyChangeEvent(this, name(), oldValue, value)
            )
    }

    override fun <T2 : Any> bindBi(
        property: CTObservableProperty<T2>,
        mapper: Function<T, T2>,
        mapper2: Function<T2, T>
    ) {
        bind(property, mapper)
        property.bind(this, mapper2)
    }

    override fun <T2 : Any> bind(property: CTProperty<T2>, mapper: Function<T, T2>) {
        val myValue = get()
        val otherValue = property.get()
        val newOtherValue = mapper.apply(myValue)
        if (otherValue != newOtherValue) {
            property.set(newOtherValue)
        }
        addPropertyChangeListener { event -> onLateBind(property, mapper, event.newValue) }
    }

    fun <T2 : Any> onLateBind(property: CTProperty<T2>, mapper: Function<T, T2>, newValue: T) {
        property.set(mapper.apply(newValue))
    }

    override fun bindBi(property: CTObservableProperty<T>) {
        bind(property)
        property.bind(this)
    }

    override fun bind(property: CTProperty<T>) {
        val myValue = get()
        val otherValue = property.get()
        if (myValue != otherValue) {
            onLateBind(property, myValue)
        }
        addPropertyChangeListener { event -> onLateBind(property, event.newValue) }
    }

    fun onLateBind(property: CTProperty<T>, newValue: T) {
        property.set(newValue)
    }

    override fun addPropertyChangeListener(listener: CTPropertyChangeListener<T>) {
        CTPropertyChangeManager.instance.addPropertyChangeListener(this, listener)
    }

    override fun removePropertyChangeListener(listener: CTPropertyChangeListener<T>) {
        CTPropertyChangeManager.instance.removePropertyChangeListener(this, listener)
    }
}
